snitun/server/worker.py

```diff
@@ -0,0 +1,121 @@
+
+        # Used on the child
+        self._peers: Optional[PeerManager] = None
+        self._list_sni: Optional[SNIProxy] = None
+        self._list_peer: Optional[PeerListener] = None
+        self._loop: Optional[asyncio.BaseEventLoop] = None
+
+        # Communication between Parent/Child
+        self._manager: Manager = Manager()
+        self._new: Queue = self._manager.Queue()
+        self._sync: Dict[str, None] = self._manager.dict()
+
+    @property
+    def peer_size(self) -> int:
+        """Return amount of managed peers."""
+        return len(self._sync)
+
+    def is_responsible_peer(self, sni: str) -> bool:
+        """Return True if worker is responsible for this peer domain."""
+        return sni in self._sync
+
+    async def _async_init(self) -> None:
+        """Initialize child process data."""
+        self._peers = PeerManager(
+            self._fernet_keys,
+            throttling=self._throttling,
+            event_callback=self._event_stream,
+        )
+        self._list_sni = SNIProxy(self._peers)
+        self._list_peer = PeerListener(self._peers)
+
+    def _event_stream(self, peer: Peer, event: PeerManagerEvent) -> None:
+        """Event stream peer connection data."""
+        if event == PeerManagerEvent.CONNECTED:
+            self._sync[peer.hostname] = None
+        else:
+            self._sync.pop(peer.hostname, None)
+
+    def shutdown(self) -> None:
+        """Shutdown child process.
+
+        This function blocking, don't call it inside loop!
+        """
+        self._new.put(None)
+        self.join(10)
+
+    def handover_connection(
+        self, con: socket, data: bytes, sni: Optional[str] = None
+    ) -> None:
+        """Move new connection to worker.
+
+        Async friendly.
+        """
+        self._new.put_nowait((con, data, sni))
+
+    def run(self) -> None:
+        """Running worker process."""
+        self._loop = asyncio.get_event_loop()
+
+        # Start eventloop
+        running_loop = Thread(target=self._loop.run_forever)
+        running_loop.start()
+
+        # Init backend
+        asyncio.run_coroutine_threadsafe(self._async_init(), loop=self._loop).result()
+
+        while True:
+            new: Tuple[socket, bytes, Optional[str]] = self._new.get()
+            if new is None:
+                break
+
+            new[0].setblocking(False)
+            asyncio.run_coroutine_threadsafe(
+                self._async_new_connection(*new), loop=self._loop
+            )
+
+        # Shutdown worker
+        self._loop.call_soon_threadsafe(self._loop.stop)
+        running_loop.join(10)
+
+    async def _async_new_connection(
+        self, con: socket, data: bytes, sni: Optional[str]
+    ) -> None:
+        """Handle incoming connection."""
+        reader, writer = await asyncio.open_connection(sock=con)
+
+        # Select the correct handler for process connection
+        if sni:
+            self._loop.create_task(
```

Comment:
Why create a task? This is already running in a task that's not awaited by `run`

Reply:
Because we don't need to track this and create a handler on the select loop if we do not need the result. After this step, the multiplexer gets over the control of the transport.